'''
Created on Nov 27, 2015

@author: Mada
'''
from Domain import Problema

class RepositoryExceptionProb(Exception):
    def __str__(self):
        return "Datele introduse nu exista."
    
class ProblemRepository:
    def __init__(self):
        self.probleme = {}
        
    def Save_Problem(self,prob):
        
        for i in self.probleme:
            if self.probleme[i].Get_Nr_Lab() == prob.Get_Nr_Lab() and self.probleme[i].Get_Nr_Prob() == prob.Get_Nr_Prob():
                raise RepositoryExceptionProb
        self.probleme[len(self.probleme)] = prob
        
    def Nr_Problems(self):
        return len(self.probleme)
    
    def Get_Problems(self):
        return list(self.probleme.values())
    
    def Find_By_Lab_Prob(self, nr_lab, nr_prob):
        for i in self.probleme:
            if self.probleme[i].Get_Nr_Lab() == nr_lab and self.probleme[i].Get_Nr_Prob() == nr_prob:
                return self.__probleme[i]
        raise RepositoryExceptionProb()
    
    def Update_Prob(self,nr_lab,nr_prob, descriere, deadline):
        ok = 0
        for i in self.probleme:
            if self.probleme[i].Get_Nr_Lab() == nr_lab and self.probleme[i].Get_Nr_Prob() == nr_prob:
                self.probleme[i].Set_Description(descriere)
                self.probleme[i].Set_Deadline(deadline)
                ok = 1
        if ok == 0:
            raise RepositoryExceptionProb()
    
    def Delete(self, nr_lab, nr_prob):
        for i in self.probleme:
            if self.probleme[i].Get_Nr_Lab() == nr_lab and self.probleme[i].Get_Nr_Prob() == nr_prob:
                self.probleme.pop(i)
                break
        else:
            raise RepositoryExceptionProb
    
    

def TestSaveProb():
    rep  = ProblemRepository()
    
    prob = Problema.Problema(4,3,'Number 1','5/5/2009')
    #assert rep.size() == 0
    try:
        rep.Save_Problem(prob)
        assert rep.Nr_Problems() == 1
    except RepositoryExceptionProb:
        print("Datele introduse deja exista.")
        
    prob1 = Problema.Problema(4,1,'Number 2','5/5/2009')
    try:
        rep.Save_Problem(prob1)
        assert rep.Nr_Problems() == 2
    except RepositoryExceptionProb:
        print("Datele introduse deja exista.")
        
def TestUpdateProb():
    rep  = ProblemRepository()
    prob = Problema.Problema(4,3,'Incearca','5/5/2009')
    '''
    Se incearca salvarea datelor(daca nu exista deja se continua)
    '''
    try:
        rep.Save_Problem(prob)
        assert rep.Nr_Problems() == 1
        '''
        Se actualizeaza datele, daca exista, in caz contrar se afiseaza mesajul corespunzator.
        '''
        try : 
            rep.Update_Prob(4,3,'Merge','6/6/2003')
        except RepositoryExceptionProb:
            print("Datele introduse nu exista, deci nu pot fi actualizate.")
        
    except RepositoryExceptionProb:
        print("Datele introduse deja exista.")
    
    
    
def TestFindProb():
    rep  = ProblemRepository()

    prob = Problema.Problema(4,3,'Prima','5/5/2009')
    rep.Save_Problem(prob)
    
    prob1 = Problema.Problema(4,1,'A doua','5/5/2009')
    rep.Save_Problem(prob)
    
    try:
        rep.Find_By_Lab_Prob(4,1)
        assert rep.Find_By_Lab_Prob(4,1) == prob1
    except RepositoryExceptionProb as e:
        print(str(e))
    

TestSaveProb()
TestUpdateProb()
TestFindProb()